docs/running-calyx/fud2/rhai-api.md

```diff
@@ -0,0 +1,152 @@
+# Rhai API
+
+## High Level Rhai
+
+### defop
+
+```
+defop <op name>(<input1>: <input1 state>, <input2>: <input2 state> ...) >> <target1>: <target1 state>, <target2>, <target2 state> ... {
+    <statements>
+}
+```
+Defines an op with `<op name>` which runs `<statements>` to generate target states from input states. 
+
+The name, inputs, and targets are called the signature of the op. `<statements>` is called the body of the op.
+
+### shell
+
+```
+shell(<string>)
+```
+When called in the body of an op, that op will run `<string>` as a shell command to generate its targets. It is an error to call `shell` outside of the body of an op. Additionally, it is an error to call `shell` in the body of an op in which `shell_deps` was called prior.
```

Comment:
Maybe somewhere we should be explicit about what the general rule is about `shell` vs. `shell_deps`? As in, you pick one or the other for a given `defop`, and you stick to that (no mixing). Normal `shell` just runs things in order, preventing all parallelism/incrementalism, and is suitable for simple recipes. `shell_deps` makes you actually declare dependencies, enabling parallelism & incremental builds.